### PR TITLE
EnGirlA

### PR DESCRIPTION
**Pull request by Kelebek1**
All hail Roman!

### CODE REVIEW

File: asm/non_matchings/overlays/actors/ovl_En_GirlA/func_80A3A8D0.s [hunk 1]
```diff
@@ -1,202 +1,4 @@
 .rdata
-glabel D_80A3CCD0
```
Comment (ethteck) [inline asm/non_matchings/overlays/actors/ovl_En_GirlA/func_80A3A8D0.s:2]: Why is some but not all of the rodata removed from this function?

File: src/overlays/actors/ovl_En_GirlA/z_en_girla.h [hunk 2]
```diff
@@ -8,7 +8,32 @@ struct EnGirlA;
 
 typedef struct EnGirlA {
     /* 0x0000 */ Actor actor;
-    /* 0x014C */ char unk_14C[0x88];
+    /* 0x014C */ SkelAnime skelAnime;
+    /* 0x0190 */ void (*actionFunc)(struct EnGirlA*, GlobalContext*);
```
Comment (ethteck) [inline src/overlays/actors/ovl_En_GirlA/z_en_girla.h:12]: A typedef should be created for the actionfunc / drawfunc formats

File: src/overlays/actors/ovl_En_GirlA/z_en_girla.c [hunk 3]
```diff
@@ -20,139 +20,986 @@ const ActorInit En_GirlA_InitVars = {
...
+    /* 0x14 */ s32 (*unk_14)(GlobalContext*, EnGirlA*);
+    /* 0x18 */ void (*unk_18)(GlobalContext*, EnGirlA*);
+    /* 0x1C */ void (*unk_1C)(GlobalContext*, EnGirlA*);
+} GirlAThing; // size = 0x20
+
+// void func_8002ED80(EnGirlA*, GlobalContext*, s32); 
+// void func_8002EBCC(EnGirlA*, GlobalContext*, s32);
+void func_80A3C498(Actor*, GlobalContext*, s32); 
+
+s32 func_80A3AAA8(GlobalContext*, EnGirlA*); s32 func_80A3AB58(GlobalContext*, EnGirlA*); s32 func_80A3ABF8(GlobalContext*, EnGirlA*);
```
Comment (ethteck) [inline src/overlays/actors/ovl_En_GirlA/z_en_girla.c:104]: These should each be on their own line. Hopefully format.sh covers it.

File: src/overlays/actors/ovl_En_GirlA/z_en_girla.c [hunk 4]
```diff
@@ -20,139 +20,986 @@ const ActorInit En_GirlA_InitVars = {
...
+        if (gSaveContext.itemGetInf[0] & 0x20) {
+            this->actor.params = 0x26;
+            return 1;
+        }
+        break;
+    }
+    return 0;
+}
+
+void func_80A3BFE4(EnGirlA*, GlobalContext*);
```
Comment (ethteck) [inline src/overlays/actors/ovl_En_GirlA/z_en_girla.c:241]: All function declarations should be with each other at the top of the file

File: src/overlays/actors/ovl_En_GirlA/z_en_girla.c [hunk 5]
```diff
@@ -20,139 +20,986 @@ const ActorInit En_GirlA_InitVars = {
...
+
+#ifdef NON_MATCHING
+void func_80A3A8D0(EnGirlA* this, GlobalContext* globalCtx) {
```
Comment (ethteck) [inline src/overlays/actors/ovl_En_GirlA/z_en_girla.c:244]: Please add a comment describing the non-matching-ness of this func

File: src/overlays/actors/ovl_En_GirlA/z_en_girla.c [hunk 6]
```diff
@@ -20,139 +20,986 @@ const ActorInit En_GirlA_InitVars = {
...
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_GirlA/func_80A3BEAC.s")
-
+#endif
+
+
+void EnGirlA_Init(Actor* thisx, GlobalContext* globalCtx) {
+    EnGirlA* this = THIS;
+    func_80A3A758(this);
+    func_80A3A8D0(this, globalCtx);
+    osSyncPrintf("%s(%2d)初期設定\n", D_80A3C590[thisx->params], thisx->params);
```
Comment (ethteck) [inline src/overlays/actors/ovl_En_GirlA/z_en_girla.c:284]: Is it possible to use this->actor.params here? I think we've agreed on using `this` for actor references whenever possible, even if we have `thisx` too. just for consistency

File: src/overlays/actors/ovl_En_GirlA/z_en_girla.c [hunk 7]
```diff
@@ -20,139 +20,986 @@ const ActorInit En_GirlA_InitVars = {
...
+void func_80A3B634(GlobalContext* globalCtx, EnGirlA* this) {
+    Inventory_ChangeAmmo(3, this->unk_1CA);
+    Rupees_ChangeBy(-this->unk_1C8);
+}
+
+
+void func_80A3B678(GlobalContext* globalCtx, EnGirlA* this) {
+    switch (this->unk_1CA) {
+    case 5:
+        Item_Give(globalCtx, 0x8E);
```
Comment (ethteck) [inline src/overlays/actors/ovl_En_GirlA/z_en_girla.c:608]: I'm pretty sure these are defined in z64item.h or something with a similar name

File: src/overlays/actors/ovl_En_GirlA/z_en_girla.c [hunk 8]
```diff
@@ -20,139 +20,986 @@ const ActorInit En_GirlA_InitVars = {
...
+        Item_Give(globalCtx, 0x91);
+        break;
+    }
+    Rupees_ChangeBy(-this->unk_1C8);
+}
+
+void func_80A3B714(GlobalContext* globalCtx, EnGirlA* this) {
+    switch (this->unk_1CA) {
+    case 5:
+        Item_Give(globalCtx, 0x8C);
```
Comment (ethteck) [inline src/overlays/actors/ovl_En_GirlA/z_en_girla.c:626]: here too

Comment (ethteck) [reply src/overlays/actors/ovl_En_GirlA/z_en_girla.c:626]: I'll stop pointing them out but all item ids

File: src/overlays/actors/ovl_En_GirlA/z_en_girla.c [hunk 9]
```diff
@@ -20,139 +20,986 @@ const ActorInit En_GirlA_InitVars = {
...
+            gSaveContext.itemGetInf[0] |= 0x20;
+            break;
+    }
+    Rupees_ChangeBy(-this->unk_1C8);
+}
+
+void func_80A3BD80(GlobalContext* globalCtx, EnGirlA* this) {
+}
+
+void func_80A3C4D4(Actor*, GlobalContext*);
```
Comment (ethteck) [inline src/overlays/actors/ovl_En_GirlA/z_en_girla.c:778]: Nyoom these to the top please

File: src/overlays/actors/ovl_En_GirlA/z_en_girla.c [hunk 10]
```diff
@@ -20,139 +20,986 @@ const ActorInit En_GirlA_InitVars = {
...
+void func_80A3BEAC(GlobalContext* globalCtx, EnGirlA* this) {
+    this->unk_1A8 = 1;
+    this->actor.draw = NULL;
+    if ((this->actor.params >= 0x1E) && (this->actor.params < 0x26)) {
+        this->actor.textId = 0xBD;
+    }
+}
+
+#ifdef NON_MATCHING
+void func_80A3BEE0(GlobalContext* globalCtx, EnGirlA* this) {
```
Comment (ethteck) [inline src/overlays/actors/ovl_En_GirlA/z_en_girla.c:838]: Could you describe the issues with this one too?

File: src/overlays/actors/ovl_En_GirlA/z_en_girla.c [hunk 11]
```diff
@@ -20,139 +20,986 @@ const ActorInit En_GirlA_InitVars = {
...
+            func_80A3BEAC(globalCtx, this);
+            return 1;
+        }
+        func_80A3BD8C(globalCtx, this);
+        return 1;
+    }
+    return 0;
+}
+
+void func_80A3C3BC(EnGirlA*, GlobalContext*);
```
Comment (ethteck) [inline src/overlays/actors/ovl_En_GirlA/z_en_girla.c:866]: ^

File: src/overlays/actors/ovl_En_GirlA/z_en_girla.c [hunk 12]
```diff
@@ -20,139 +20,986 @@ const ActorInit En_GirlA_InitVars = {
...
+    func_80A3BF54(this, globalCtx);
+    this->actionFunc(this, globalCtx);
+    Actor_SetHeight(&this->actor, 5.0f);
+    this->actor.shape.rot.x = 0.0f;
+    if (this->actor.params != 0x26) {
+        if (this->unk_1B4 != 0) {
+            this->unk_1B8 += 0x1F4;
+            return;
+        }
+        Math_SmoothScaleMaxMinS(&this->unk_1B8, 0, 10, 0x7D0, 0);
```
Comment (ethteck) [inline src/overlays/actors/ovl_En_GirlA/z_en_girla.c:984]: I'd make this an if-else

File: asm/non_matchings/overlays/actors/ovl_En_GirlA/func_80A3A8D0.s (hunk 1 above)
Comment (Kelebek1) [reply asm/non_matchings/overlays/actors/ovl_En_GirlA/func_80A3A8D0.s:2]: I have to do it for the data to match, I don't understand why though. Fig helped me with it.

File: src/overlays/actors/ovl_En_GirlA/z_en_girla.c [hunk 13]
```diff
@@ -1,14 +1,68 @@
...
 #define FLAGS 0x00000019
 
 #define THIS ((EnGirlA*)thisx)
 
-void EnGirlA_Init(Actor* thisx, GlobalContext* globalCtx);
-void EnGirlA_Destroy(Actor* thisx, GlobalContext* globalCtx);
-void EnGirlA_Update(Actor* thisx, GlobalContext* globalCtx);
+void EnGirlA_Init(Actor* this, GlobalContext* globalCtx);
+void EnGirlA_Destroy(Actor* this, GlobalContext* globalCtx);
+void EnGirlA_Update(Actor* this, GlobalContext* globalCtx);
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_GirlA/z_en_girla.c:10]:
```suggestion
void EnGirlA_Init(Actor* thisx, GlobalContext* globalCtx);
void EnGirlA_Destroy(Actor* thisx, GlobalContext* globalCtx);
void EnGirlA_Update(Actor* thisx, GlobalContext* globalCtx);
```

File: src/overlays/actors/ovl_En_GirlA/z_en_girla.c [hunk 14]
```diff
@@ -1,14 +1,68 @@
...
+
+void func_80A3BEAC(GlobalContext*, EnGirlA*);
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_GirlA/z_en_girla.c:12]: Ideally, forward declarations like these should have argument names (ie they can just be a copy paste of the actual function header)

File: src/overlays/actors/ovl_En_GirlA/z_en_girla.c [hunk 15]
```diff
@@ -20,139 +74,901 @@ const ActorInit En_GirlA_InitVars = {
...
+    "鶏の卵        ", "インゴー牛乳  ", "インゴー卵    ", "もだえ石      ", "大人の財布    ", "ハートの欠片  ",
+    "ボムチュウ    ", "ボムチュウ    ", "ボムチュウ    ", "ボムチュウ    ", "ボムチュウ    ", "デクのタネ    ",
+    "キータンのお面", "こわそなお面  ", "ドクロのお面  ", "ウサギずきん  ", "まことの仮面  ", "ゾーラのお面  ",
+    "ゴロンのお面  ", "ゲルドのお面  ", "ＳＯＬＤＯＵＴ", "炎            ", "虫            ", "チョウチョ    ",
+    "ポウ          ", "妖精の魂      ", "矢×10        ",  "爆弾×20      ",  "爆弾×30      ",  "爆弾×5       ",
+    "赤クスリ      ", "赤クスリ      ",
+};
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_GirlA/func_80A3A8A0.s")
+s16 D_80A3C658[8] = { 0x0024, 0x0026, 0x0025, 0x0027, 0x002B, 0x0029, 0x0028, 0x002A };
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_GirlA/z_en_girla.c:90]: Based on its usage, this appears to be an array of item IDs so its value can be values from the corresponding enum (`ITEM_*`)

File: src/overlays/actors/ovl_En_GirlA/z_en_girla.c [hunk 16]
```diff
@@ -20,139 +74,901 @@ const ActorInit En_GirlA_InitVars = {
...
+    /* 0x04 */ void (*unk_04)(Actor*, GlobalContext*, s32);
+    /* 0x08 */ s16 unk_08;
+    /* 0x0A */ s16 unk_0A;
+    /* 0x0C */ u16 unk_0C;
+    /* 0x0C */ u16 unk_0E;
+    /* 0x10 */ s32 unk_10;
+    /* 0x14 */ s32 (*unk_14)(GlobalContext*, EnGirlA*);
+    /* 0x18 */ void (*unk_18)(GlobalContext*, EnGirlA*);
+    /* 0x1C */ void (*unk_1C)(GlobalContext*, EnGirlA*);
+} GirlAThing; // size = 0x20
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_GirlA/z_en_girla.c:106]:
```suggestion
} ShopItemEntry; // size = 0x20
```

File: src/overlays/actors/ovl_En_GirlA/z_en_girla.c [hunk 17]
```diff
@@ -20,139 +74,901 @@ const ActorInit En_GirlA_InitVars = {
...
+
+GirlAThing D_80A3C674[] = {
+    { 0x00BB, 0x0011, func_8002ED80, 0x000F, 0x0005, 0x00B2, 0x007F, GI_NUTS_5_2, func_80A3ABF8, func_80A3B714, func_80A3BB6C },
+    { 0x00D8, 0x0025, func_8002EBCC, 0x003C, 0x001E, 0x00C1, 0x009B, GI_ARROWS_MEDIUM, func_80A3AAA8, func_80A3B634, func_80A3BB6C },
+    { 0x00D8, 0x0026, func_8002EBCC, 0x005A, 0x0032, 0x00B0, 0x007D, GI_ARROWS_LARGE, func_80A3AAA8, func_80A3B634, func_80A3BB6C },
+    { 0x00CE, 0x001F, func_8002EBCC, 0x0019, 0x0005, 0x00A3, 0x008B, GI_BOMBS_5, func_80A3AB58, func_80A3B678, func_80A3BB6C },
+    { 0x00BB, 0x0011, func_8002ED80, 0x001E, 0x000A, 0x00A2, 0x0087, GI_NUTS_10, func_80A3ABF8, func_80A3B714, func_80A3BB6C },
+    { 0x00C7, 0x001A, 0x00000000, 0x000A, 0x0001, 0x00A1, 0x0088, GI_STICKS_1, func_80A3ACAC, func_80A3B780, func_80A3BB6C },
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_GirlA/z_en_girla.c:114]: 0 words in places where there are normally pointers should be `NULL`, like here
```suggestion
    { 0x00C7, 0x001A, NULL, 0x000A, 0x0001, 0x00A1, 0x0088, GI_STICKS_1, func_80A3ACAC, func_80A3B780, func_80A3BB6C },
```

File: src/overlays/actors/ovl_En_GirlA/z_en_girla.c [hunk 18]
```diff
@@ -20,139 +74,901 @@ const ActorInit En_GirlA_InitVars = {
...
+}
+
+#ifdef NON_MATCHING
+// Weird allocations
+void func_80A3A8D0(EnGirlA* this, GlobalContext* globalCtx) {
+    char** strptr = &D_80A3C590[this->actor.params];
+    s16 params = this->actor.params;
+
+    if (strptr) {};
+    osSyncPrintf("%s(%2d)初期設定\n", *(&D_80A3C590[this->actor.params]), this->actor.params);
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_GirlA/z_en_girla.c:234]:
```suggestion
    osSyncPrintf("%s(%2d)初期設定\n", D_80A3C590[this->actor.params], this->actor.params);
```

File: src/overlays/actors/ovl_En_GirlA/z_en_girla.c [hunk 19]
```diff
@@ -20,139 +74,901 @@ const ActorInit En_GirlA_InitVars = {
...
+    if ((params >= 50) && (params < 0)) {
+        Actor_Kill(&this->actor);
+        osSyncPrintf(VT_COL(RED, WHITE));
+        osSyncPrintf("引数がおかしいよ(arg_data=%d)！！\n", this->actor.params);
+        osSyncPrintf(VT_RST);
+        if (!this->actor.params) {};
+        __assert("0", "../z_en_girlA.c", 0x58D);
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_GirlA/z_en_girla.c:241]: The third argument in __assert calls is a line number so they should be decimal ideally.

File: src/overlays/actors/ovl_En_GirlA/z_en_girla.c [hunk 20]
```diff
@@ -20,139 +74,901 @@ const ActorInit En_GirlA_InitVars = {
...
+}
+
+s32 func_80A3ACAC(GlobalContext* globalCtx, EnGirlA* this) {
+    if ((CUR_CAPACITY(UPG_STICKS) != 0) && (AMMO(ITEM_STICK) >= CUR_CAPACITY(UPG_STICKS))) {
+        return 2;
+    }
+    if (gSaveContext.rupees < this->unk_1C8) {
+        return 4;
+    }
+    if (Item_CheckObtainability(ITEM_STICK) == 0xFF) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_GirlA/z_en_girla.c:324]:
```suggestion
    if (Item_CheckObtainability(ITEM_STICK) == ITEM_NONE) {
```

File: src/overlays/actors/ovl_En_GirlA/z_en_girla.c [hunk 21]
```diff
@@ -20,139 +74,901 @@ const ActorInit En_GirlA_InitVars = {
...
+        return 4;
+    }
+    if (Item_CheckObtainability(ITEM_FAIRY) == ITEM_NONE) {
+        return 0;
+    }
+    return 1;
+}
+
+void func_80A3B634(GlobalContext* globalCtx, EnGirlA* this) {
+    Inventory_ChangeAmmo(3, this->unk_1CA);
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_GirlA/z_en_girla.c:574]:
```suggestion
    Inventory_ChangeAmmo(ITEM_BOW, this->unk_1CA);
```

File: src/overlays/actors/ovl_En_GirlA/z_en_girla.c [hunk 22]
```diff
@@ -20,139 +74,901 @@ const ActorInit En_GirlA_InitVars = {
...
+            break;
+        case 10:
+            Item_Give(globalCtx, ITEM_NUTS_10);
+            break;
+    }
+    Rupees_ChangeBy(-this->unk_1C8);
+}
+
+void func_80A3B780(GlobalContext* globalCtx, EnGirlA* this) {
+    Item_Give(globalCtx, 0);
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_GirlA/z_en_girla.c:609]:
```suggestion
    Item_Give(globalCtx, ITEM_STICK);
```

File: src/overlays/actors/ovl_En_GirlA/z_en_girla.c [hunk 23]
```diff
@@ -20,139 +74,901 @@ const ActorInit En_GirlA_InitVars = {
...
+        case 0x2B:
+            Item_Give(globalCtx, ITEM_FAIRY);
+            break;
+    }
+    Rupees_ChangeBy(-this->unk_1C8);
+}
+
+void func_80A3BB6C(GlobalContext* globalCtx, EnGirlA* this) {
+    if (this->actor.params == 0xC) {
+        if ((gSaveContext.infTable[7] & 0x40)) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_GirlA/z_en_girla.c:702]:
```suggestion
        if (gSaveContext.infTable[7] & 0x40) {
```

File: src/overlays/actors/ovl_En_GirlA/z_en_girla.c [hunk 24]
```diff
@@ -20,139 +74,901 @@ const ActorInit En_GirlA_InitVars = {
...
+    s32 params = this->actor.params;
+    s32 params2;
+    s32 tmp3;
+
+    if ((this->actor.params >= 0x1E) && (this->actor.params < 0x23)) {
+        params2 = this->actor.params - 0x1E;
+        tmp3 = 0;
+        switch (this->actor.params) {
+            case 0x1E:
+                if ((gSaveContext.itemGetInf[3] & 0x100) != 0) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_GirlA/z_en_girla.c:762]:
```suggestion
                if (gSaveContext.itemGetInf[3] & 0x100) {
```

File: src/overlays/actors/ovl_En_GirlA/z_en_girla.c [hunk 25]
```diff
@@ -20,139 +74,901 @@ const ActorInit En_GirlA_InitVars = {
...
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_GirlA/func_80A3C4D4.s")
+#endif
+
+s32 func_80A3BF54(EnGirlA* this, GlobalContext* globalCtx) {
+    s32 params;
+
+    if ((this->actor.params >= 0x1E) && (this->actor.params < 0x26)) {
+        params = this->actor.params - 0x1E;
+        if (gSaveContext.items[gItemSlots[0x21]] == D_80A3C658[params]) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_GirlA/z_en_girla.c:830]:
```suggestion
        if (INV_CONTENT(ITEM_WEIRD_EGG) == D_80A3C658[params]) {
```

File: src/overlays/actors/ovl_En_GirlA/z_en_girla.c (hunk 18 above)
Comment (Kelebek1) [reply src/overlays/actors/ovl_En_GirlA/z_en_girla.c:286]: I'm leaving this as-is, because that's the closest I can get to matching that function.

**Pull request merged**